FiltersEngine.cpp

```diff
@@ -0,0 +1,201 @@
+/*
+    microsoft-oms-auditd-plugin
+
+    Copyright (c) Microsoft Corporation
+
+    All rights reserved. 
+
+    MIT License
+
+    Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the ""Software""), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+
+    The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+#include "FiltersEngine.h"
+
+#include "Logger.h"
+
+#include <string>
+#include <iostream>
+#include <fstream>
+#include <sys/stat.h> /* for stat() */
+#include <dirent.h>
+#include <algorithm>
+#include <unistd.h>
+#include <limits.h>
+
+std::bitset<FILTER_BITSET_SIZE> FiltersEngine::AddFilter(ProcFilterSpec& pfs, std::string& outputName)
+{
+    std::bitset<FILTER_BITSET_SIZE> ret;
+
+    if (_filtersBitPosition.count(pfs) > 0) {
```

Comment:
Since the same key lookup occurs 2-3 times in this code, consider using _filtersBitPosition.find() instead (e.g. "auto itr = _filtersBitPosition.find(pfs)", then if itr != _filtersBitPosition.end(), itr->outputs and itr->bitPosition are available).

Reply:
Thanks. If loading filters becomes a performance consideration then I'll implement this.

Reply:
Accepted. I've replaced this construction with that suggested in both FiltersEngine.cpp and ProcessTree.cpp.